backend/uclapi/workspaces/occupeye/archive.py

```diff
@@ -0,0 +1,217 @@
+                # We want to keep the data with the greatest hardware ID or survey device ID
+                unique_sensors = {}
+                for key, sensor in enumerate(sensors):
+                    unique_key = str(sensor["SensorID"]) + ":" + sensor["TriggerDate"]
+                    unique_value = sensor["HardwareID"] + sensor["SurveyDeviceID"]
+                    if unique_key in unique_sensors:
+                        if unique_value > unique_sensors[unique_key][1]:
+                            unique_sensors[unique_key] = (key, unique_value)
+                    else:
+                        unique_sensors[unique_key] = (key, unique_value)
+
+                filtered_sensors = []
+                for _, value in unique_sensors.items():
+                    filtered_sensors.append(sensors[value[0]])
+
+                sensors = filtered_sensors
+                batch_objects = []
+
+                for sensor in sensors:
+                    cs_key = str(sensor["SensorID"]) + f"-{survey.survey_id}"
+                    if cs_key in self.current_sensors:
+                        self.validate_sensor(sensor, survey, start_date, end_date)
+                    else:
+                        Sensors.objects.create(sensor_id=sensor["SensorID"], hardware_id=sensor["HardwareID"],
+                                               survey_device_id=sensor["SurveyDeviceID"],
+                                               survey_id=survey.survey_id)
+                        self.current_sensors[str(sensor["SensorID"]) + "-" + str(survey.survey_id)] = {
+                            "hardware_id": sensor["HardwareID"], "survey_device_id": sensor["SurveyDeviceID"],
+                            "survey_id": survey.survey_id, "last_value": -1}
+
+                    for key, value in sensor.items():
+                        if key.startswith("t"):
+                            # A few time stamps have None instead of -1
+                            if value is None:
+                                value = -1
+                            if value != self.current_sensors[cs_key]["last_value"]:
+                                updates += 1
+                                time = datetime.strptime(sensor["TriggerDate"] + key, "%Y-%m-%dt%H%M")
+                                time = time.replace(tzinfo=timezone.utc)
+                                batch_objects.append(
+                                    Historical(sensor_id=sensor["SensorID"], survey_id=survey.survey_id,
+                                               datetime=time, state=value))
+                                self.current_sensors[cs_key]["last_value"] = value
+
+                # Ideally ignore conflicts would be false but depending on exactly when the archive is run or
+                # if the previous run crashed some conflicts may exist.
+                Historical.objects.bulk_create(batch_objects, batch_size=65536, ignore_conflicts=True)
```

Comment:
It could go outside the inner `for` loop (but not the outer as it is assumed that the data is saved). However I am not sure it is needed as bulk_create has very limited speed up when inserting over the batch_size vs running bulk_create multiple times. It is also good to wait a few seconds before requesting more data from the endpoint and we don't need to try and store gigabytes of historical data in ram.